Add database PSQL, Redis, Docker

DaryaSerova opened the pull request:
feat: добавить кэширование в Redis.
feat: создать БД в PostgreSQL, настроить подключение к БД
feat: упаковать в три контейнера - основной сервис, Redis, база данных

DaryaSerova committed:
feat: добавить кэширование в Redis.
feat: создать БД в PostgreSQL, настроить подключение к БД
feat: упаковать в три контейнера - основной сервис, Redis, база данных

Solntmore commented on pom.xml:
```diff
@@ -73,6 +73,29 @@
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-thymeleaf</artifactId>
+        </dependency>
```
а thymeleaf для чего? У нас же есть самостоятельный фронтенд)

Solntmore commented on src/main/java/ru/mom/remembers/config/RedisConfig.java:
```diff
@@ -0,0 +1,69 @@
+        RedisCacheManager redisCacheManager = RedisCacheManager.builder(factory).cacheDefaults(redisCacheConfiguration)
+                .build();
+        return redisCacheManager;
+    }
```
Можно сразу сократить до return RedisCacheManager.builder(factory).cacheDefaults(redisCacheConfiguration)
.initialCacheNames(Set.of("cityByIdCache", "citiesListCache", "attractionsListCache"))
.build();

Solntmore commented on src/main/java/ru/mom/remembers/note/service/NoteServiceImpl.java:
```diff
@@ -63,8 +76,10 @@ public FullResponseNoteDto createNote(NewRequestNoteDto newNoteDto) {
     }
 
     @Override
+    @CacheEvict(cacheNames = NOTE_CACHE,  key = "#id")
```
@CachePut если использовать, то можно не удалять кеш при обновлении в бд, а обновлять)

Solntmore commented on docker-compose.yml:
```diff
@@ -0,0 +1,38 @@
+      - POSTGRES_USER=admin
+      - POSTGRES_PASSWORD=admin
+      - POSTGRES_DB=mom-remembers
+
```
Лучше здесь в enviroment добавить       
- DB_HOST=attractionDB
- DB_PORT=5432
а в application.properties  использовать по подобному шаблону
spring.datasource.url=jdbc:postgresql://${DB_HOST:attraction-db}:${DB_PORT:5432}/${DB_NAME:attractionDB}
spring.datasource.username=${POSTGRES_USER:root}
spring.datasource.password=${POSTGRES_PASSWORD:root}
spring.cache.redis.time-to-live=30S
spring.redis.host=${SPRING_REDIS_HOST:localhost}
spring.redis.port=${SPRING_REDIS_PORT:6379}
Такая конфигурация сначала проверяет информацию из docker-compose, а затем, если там нет, ставит значение по умолчанию, которое тоже ты задаешь. Позволяет более гибко настраивать проект)

Solntmore commented on src/main/resources/application.properties:
```diff
@@ -1 +1,20 @@
+logging.level.org.springframework.transaction.interceptor=TRACE
+logging.level.org.springframework.orm.jpa.JpaTransactionManager=DEBUG
+logging.level.org.springframework.web.client.RestTemplate=DEBUG
+spring.datasource.url=jdbc:postgresql://localhost:5432/mom-remembers
```
Можно использовать такую схему, то, что выше писал в docker-compose коммент
spring.datasource.url=jdbc:postgresql://${DB_HOST:attraction-db}:${DB_PORT:5432}/${DB_NAME:attractionDB}
spring.datasource.username=${POSTGRES_USER:root}
spring.datasource.password=${POSTGRES_PASSWORD:root}
spring.cache.redis.time-to-live=30S
spring.redis.host=${SPRING_REDIS_HOST:localhost}
spring.redis.port=${SPRING_REDIS_PORT:6379}
Такая конфигурация сначала проверяет информацию из docker-compose, а затем, если там нет, ставит значение по умолчанию, которое тоже ты задаешь. Позволяет более гибко настраивать проект)

Solntmore commented on src/main/resources/application.properties:
```diff
@@ -1 +1,20 @@
+spring.main.allow-circular-references=true
+
+spring.redis.host=localhost
+spring.redis.port=6379
```
Думаю, стоит добавить конфигурацию для запуска не в контейнерах, а локально тоже) Иначе как оперативно тестировать, каждый раз контейнер перезаписывать?
Это может выглядеть так(Кстати, там же можешь взять всякие полезные конфигурации для кеша в редиса, например время жизни кеша spring.cache.redis.time-to-live=30S, которое защищает нас от переполнения)
#---
spring.config.activate.on-profile=ci,test,dev
spring.datasource.driverClassName=org.h2.Driver
spring.datasource.url=jdbc:h2:mem:ewm
spring.datasource.username=${POSTGRES_USER:root}
spring.datasource.password=${POSTGRES_PASSWORD:root}
spring.cache.redis.time-to-live=30S
spring.redis.host=localhost
spring.redis.port=6379
spring.redis.timeout=10000ms
spring.redis.lettuce.pool.max-active=8
spring.redis.lettuce.pool.max-wait=-1ms
spring.redis.lettuce.pool.max-idle=8
spring.redis.lettuce.pool.min-idle=8
spring.cache.type=redis
spring.main.allow-circular-references=true
#---

Solntmore commented on Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM amazoncorretto:11
+COPY /target/redis-cache-0.0.1-SNAPSHOT.jar redis-cache-0.0.1-SNAPSHOT.jar
```
При выполнении package формируется такой jar remembers-0.0.1-SNAPSHOT.jar, а тут другое название, из-за этого доккер не видит образ и не может собрать контейнер) Для redis вообще не нужен dockerfile, но я думаю, ты просто опечаталась. Тут название jar нашего основного сервиса.

Solntmore commented on src/main/java/ru/mom/remembers/note/service/NoteServiceImpl.java:
```diff
@@ -36,6 +48,7 @@ public class NoteServiceImpl implements NoteService {
     private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
 
     @Override
+    @Cacheable(cacheNames = NOTE_CACHE)
```
если не указываем другие поля типо key или еще что-то, можно вообще cacheNames = убрать и оставить просто название кеша, в нашем случае NOTE_CACHE

Solntmore commented on src/main/java/ru/mom/remembers/note/service/NoteServiceImpl.java:
```diff
@@ -94,8 +110,10 @@ public void deleteNote(Long id) {
         notePersistService.deleteNote(id);
     }
 
+    @Cacheable(cacheNames = NOTE_CACHE, key = "#id")
```
Так работать, к сожалению, не будет, о чем и идея сигнализирует
Правильная сигнатура такая
 @Cacheable(cacheNames = {NOTE_CACHE}, key = "#id")
Но, в целом это здесь и не нужно, приватные методы не кешируетчя, об этом, кстати, тоже идея сигнализирует и просить сделать метод не приватным)

DaryaSerova committed:
fix: рефакторинг RedisConfig
fix: правка аннотаций Cacheable
feat: добавить application-test.properties для локального запуска в режиме test
fix: правка docker-compose.yml, Dockerfile

Solntmore commented on docker-compose.yml:
```diff
@@ -3,23 +3,25 @@ services:
+    container_name: mom-remembers-main
     ports:
-      - "8080:8080"
+      - "9090:9090"
```
а зачем на порт 9090 переехали?)

Solntmore replied:
а в application 8090 порт вообще указан)

Solntmore commented on docker-compose.yml:
```diff
@@ -3,23 +3,25 @@ services:
   mom-remembers-db:
     image: postgres:14-alpine
-    container_name: mom-remembers-db
+    container_name: mom-remembers
```
имя контейнера лучше, чтоб содержала DB, если это контейнер базы данных, так в Docker гораздо понятнее, какой контейнер за что отвечает

Solntmore commented on src/main/resources/application-test.properties:
```diff
@@ -0,0 +1,16 @@
+server.port=8090
```
так в итоге 8080, 9090 или 8090?) Можно main server 8080 сделать во всех местах и бед не знать)
